bugsnag-android-core/src/main/java/com/bugsnag/android/DocumentPath.java

```diff
@@ -0,0 +1,385 @@
+                    break;
+            }
+        }
+        if (strBegin < path.length()) {
+            String pathComponent = null;
+            if (requiresEscaping) {
+                pathComponent = unescape(path, strBegin, path.length());
+            } else {
+                pathComponent = path.substring(strBegin);
+            }
+            directives.add(makeDirectiveFromPathComponent(pathComponent));
+        }
+
+        if (isDotLast(path)) {
+            directives.add(new Directive.ListInsertDirective());
+        }
+
+        return directives;
+    }
+
+    private static String unescape(String str, int startOffset, int endOffset) {
+        StringBuilder buff = new StringBuilder(endOffset - startOffset);
+        for (int i = startOffset; i < endOffset; i++) {
+            char ch = str.charAt(i);
+            if (ch == '\\') {
+                i++;
+                if (i >= endOffset) {
+                    throw new IllegalArgumentException("Path cannot end on an escape char '\\'");
+                }
+                buff.append(str.charAt(i));
+            } else {
+                buff.append(ch);
+            }
+        }
+        return buff.toString();
+    }
+
+    private static boolean isDotLast(String path) {
+        if (path.charAt(path.length() - 1) != '.') {
+            return false;
+        }
+
+        if (path.length() == 1) {
+            return true;
+        }
+
+        boolean isUnescapedDot = true;
+        for (int i = path.length() - 2; i >= 0; i--) {
+            char ch = path.charAt(i);
+            if (ch == '\\') {
+                isUnescapedDot = !isUnescapedDot;
+            } else {
+                break;
+            }
+        }
+
+        return isUnescapedDot;
+    }
+
+    private static Directive makeDirectiveFromPathComponent(String pathComponent) {
+        if (isInteger(pathComponent)) {
+            int index = Integer.parseInt(pathComponent);
+            if (index == -1) {
+                return new Directive.ListLastIndexDirective();
+            } else {
+                return new Directive.ListIndexDirective(index);
+            }
+        }
+        return new Directive.MapKeyDirective(pathComponent);
+    }
+
+    private static boolean isInteger(String str) {
+        if (str.length() == 0) {
+            return false;
+        }
+
+        int startIndex = 0;
+        if (str.charAt(0) == '-') {
+            startIndex = 1;
+        }
+
+        for (int i = startIndex; i < str.length(); i++) {
+            int ch = str.charAt(i);
+            if (ch < '0' || ch > '9') {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * A directive determines how document path commands are interpreted. Depending on the
+     * current context, making a container, getting from a container, or setting to a container
+     * will require different operations to complete.
+     */
+    private interface Directive {
```

Comment:
Consider moving this to a separate source file as the file is currently >300 lines, and will probably grow over time